### PR TITLE
Validate config file against pipeline.

### PR DESCRIPTION
**Pull request by banasraf**
Most of the volume of this PR is made by config files for test models - they're mostly just copy-paste with slight modifications (to introduce validation errors).

The significant parts are in files: 
- `dali_pipeline.h/cc` - methods for fetching inputs and outputs info from the pipeline.
- `dali_model.h` - instantiating pipeline (with choosing a device for that) and calling actual validation.

Tests are implemented in `client.py`. They try to load each of the prepared models and expect some kind of error.

### CODE REVIEW

File: src/dali_model.cc [hunk 1]
```diff
@@ -25,16 +25,17 @@
 namespace triton { namespace backend { namespace dali {
 
 TRITONSERVER_Error* DaliModel::Create(TRITONBACKEND_Model* triton_model, DaliModel** state) {
-  TRITONSERVER_Error* error = nullptr;  // success
   try {
     *state = new DaliModel(triton_model);
+    return nullptr;
+  } catch (TritonError& e) {
+    return e.release();
```
Comment (jantonguirao) [inline src/dali_model.cc:32]: who deletes this?

File: src/dali_model.h [hunk 2]
```diff
@@ -184,10 +192,85 @@ class DaliModel : public ::triton::backend::BackendModel {
     return ret.empty() ? "model.dali" : ret;
   }
 
+  /**
+   * @brief Returns a device id on which DALI pipeline can be instantiated,
+   *   or CPU_ONLY_DEVICE_ID if only cpu is available
+   */
+  int FindFittingDevice() {
```
Comment (jantonguirao) [inline src/dali_model.h:199]:
```suggestion
  int FindDevice() {
```

**Review by jantonguirao — APPROVED** [review]

File: src/dali_model.h (hunk 2 above)
Comment (banasraf) [reply src/dali_model.h]: done

File: src/dali_model.cc (hunk 1 above)
Comment (banasraf) [reply src/dali_model.cc:32]: Triton server. The error is returned and propagated up to the server which handles it.

File: src/dali_executor/dali_pipeline.cc [hunk 3]
```diff
@@ -95,10 +95,54 @@ void DaliPipeline::PutOutput(void* destination, int output_idx, device_type_t de
...
+  }
+  return result;
+}
+
+std::optional<std::vector<int64_t>> DaliPipeline::GetInputShape(const std::string &name) {
+  int ndim = daliGetExternalInputNdim(&handle_, name.c_str());
+  if (ndim >= 0) {
+    return {std::vector<int64_t>(ndim, -1)};
+  } else {
+    return {};
```
[the remark this answers is not in the document]
Comment (banasraf) [reply src/dali_executor/dali_pipeline.cc]: done

File: src/dali_executor/dali_pipeline.cc [hunk 4]
```diff
@@ -95,10 +95,54 @@ void DaliPipeline::PutOutput(void* destination, int output_idx, device_type_t de
...
 
+
+std::string DaliPipeline::GetOutputName(int id) {
+  return std::string(daliGetOutputName(&handle_, id));
+}
+
+std::optional<std::vector<int64_t>> DaliPipeline::GetDeclaredOutputShape(int id) {
+  int ndim = daliGetDeclaredOutputNdim(&handle_, id);
+  if (ndim >= 0) {
+    return {std::vector<int64_t>(ndim, -1)};
```
[the remark this answers is not in the document]
Comment (banasraf) [reply src/dali_executor/dali_pipeline.cc]: done

File: src/dali_executor/dali_pipeline.cc [hunk 5]
```diff
@@ -95,10 +95,54 @@ void DaliPipeline::PutOutput(void* destination, int output_idx, device_type_t de
...
+  for (int i = 0; i < num_inputs; ++i) {
+    result[i] = daliGetExternalInputName(&handle_, i);
+  }
+  return result;
+}
+
+std::optional<std::vector<int64_t>> DaliPipeline::GetInputShape(const std::string &name) {
+  int ndim = daliGetExternalInputNdim(&handle_, name.c_str());
+  if (ndim >= 0) {
+    return {std::vector<int64_t>(ndim, -1)};
```
[the remark this answers is not in the document]
Comment (banasraf) [reply src/dali_executor/dali_pipeline.cc]: done

File: qa/L0_config_validation/client.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+def assert_error(func, *args, contains=None):
+  try:
+    func(*args)
+    msg = "Should raise error: " + ', '.join([str(arg) for arg in args])
+    assert False, msg
+  except InferenceServerException as err:
+    err_msg = str(err)
+  if contains is not None:
+    assert contains in err_msg, 'Error message:\n' + err_msg + '\nshould contain:\n' + contains
```
[the remark this answers is not in the document]
Comment (banasraf) [reply qa/L0_config_validation/client.py]: done

**Review by prak-nv — APPROVED** [review]

**Pull request merged**